src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -0,0 +1,536 @@
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(latestNewHeader), latestNewHeader);
+
+            ConnectedHeaders connectedHeaders = new ConnectedHeaders();
+            connectedHeaders.DownloadTo = connectedHeaders.Consumed = latestNewHeader;
+
+            ChainedHeader current = latestNewHeader;
+            ChainedHeader next = current;
+
+            while (this.HeaderWasNotRequested(current))
+            {
+                current.BlockDataAvailability = BlockDataAvailabilityState.BlockRequired;
+
+                next = current;
+                current = current.Previous;
+            }
+
+            connectedHeaders.DownloadFrom = next;
+
+            this.logger.LogTrace("(-):{0}", connectedHeaders);
+            return connectedHeaders;
+        }
+
+        /// <summary>
+        /// Mark all previous blocks to <see cref="chainedHeader"/> as <see cref="ValidationState.AssumedValid"/>.
+        /// </summary>
+        /// <param name="chainedHeader">The new header that represents a longer chain.</param>
+        private void MarkTrustedChainAsAssumedValid(ChainedHeader chainedHeader)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(chainedHeader), chainedHeader);
+
+            ChainedHeader current = chainedHeader;
+
+            while (this.HeaderWasNotMarkedAsValidated(current))
+            {
+                current.BlockValidationState = ValidationState.AssumedValid;
+                current = current.Previous;
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// The header is assumed to be valid, the header and all of its previous headers will be marked as <see cref="ValidationState.AssumedValid"/>.
+        /// If the header's cumulative work is better then <see cref="IChainState.ConsensusTip"/> the header and all its predecessors will be marked with <see cref="BlockDataAvailabilityState.BlockRequired"/>.
+        /// </summary>
+        /// <param name="assumedValidHeader">The header that is assumed to be valid.</param>
+        /// <param name="latestNewHeader">The last header in the list of presented new headers.</param>
+        private ConnectedHeaders HandleAssumedValidHeader(ChainedHeader assumedValidHeader, ChainedHeader latestNewHeader)
+        {
+            this.logger.LogTrace("({0}:'{1}',{2}:'{3}')", nameof(assumedValidHeader), assumedValidHeader, nameof(latestNewHeader), latestNewHeader);
+
+            ChainedHeader bestTip = this.chainState.ConsensusTip;
+            ConnectedHeaders connectedHeaders = new ConnectedHeaders() {Consumed = latestNewHeader};
+
+            if (latestNewHeader.ChainWork > bestTip.ChainWork)
+            {
+                this.logger.LogDebug("Chained header '{0}' is the tip of a chain with more work than our current consensus tip.", latestNewHeader);
+
+                connectedHeaders = this.MarkBetterChainAsRequired(latestNewHeader);
```

Comment:
Here we must distinguish between 2 situations: 1 –  we ARE before last checkpoint, in which case we only want to download up to assume valid itself. 2 – we are NOT before last checkpoint, in which case we want really to download everything up to the last new header.